src/test/io/github/christophermanahan/carnitas/EchoServerTest.java

```diff
@@ -0,0 +1,99 @@
+package io.github.christophermanahan.carnitas;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class EchoServerTest {
+
+  private List<String> incomingData;
+  private TestSocket socket;
+
+  @BeforeEach
+  void setup() throws IOException {
+    incomingData = List.of("data", "data");
+    socket = new TestSocket(incomingData);
+
+    new EchoServer(new TestServerSocket(socket)).run();
+  }
+
+  @Test
+  void dataReceivedFromInputIsSentToOutputUntilDisconnect() throws IOException {
+    String outgoingData = socket.getOutputStream().toString();
+    List<String> outgoingDataList = List.of(outgoingData.split("\n"));
+    assertEquals(incomingData, outgoingDataList);
+  }
+
+  @Test void socketIsClosedWhenStreamIsDisconnected() throws IOException {
+    assertTrue(socket.closed);
+  }
```

Comment:
I half expected there to be more than one test asserting that data that is received is sent back. One a second pass, I can see that this is implicitly tested via checking that socket is closed.

I am wondering if we are leaning too much on the behavior of the test double. It is pretty complicated. Does it needs tests as well?

(I intended to leave a form of this comment on the initial review, but it looks like it got lost somewhere)

Reply:
I do believe the test double is doing a lot of the heavy lifting. However, I was having trouble driving the design of the HTTP server without the TestSocket emulating a lot of the behavior of a real socket using ByteArrayStreams. I think I can simplify the test double significantly by utilizing some injected class that can get the socket connection for me and return a connection class that I control and can therefore mock (instead of newing up the connection in the EchoServer and passing in this complex TestSocket). But that would require me to perform the Connection refactor so that I can return a single class that has the ability to open, close, send, and receive (and therefore mock all of those behaviors).

Reply:
One current benefit though is that mocking these streams helped me to understand the socket interface and how it functions!

Reply:
>But that would require me to perform the Connection refactor so that I can return a single class that has the ability to open, close, send, and receive (and therefore mock all of those behaviors)

Sounds good. What I am hearing is that we are choosing to take on some technical debt. I would encourage you to factor that into the what is committed to in the next iteration.

Reply:
That makes sense. Thank you for pointing that out I will keep it in mind during the IPM.